README.md
# Singularity Container Tools

[![PyPI version](https://badge.fury.io/py/stools.svg)](https://badge.fury.io/py/stools)

These are tools for Singularity containers, optimized for using with continuous integration for security
and quality checks. For an example of the package being used in a continuous integration context, see the [stools-clair](https://github.com/singularityhub/stools-clair) repository.

[![asciicast](https://asciinema.org/a/178712.png)](https://asciinema.org/a/178712)

In this work we will use [Clair OS](https://github.com/coreos/clair) combined with Continuous Integration
(travis and circle) to scan [Singularity](https://singularityware.github.io) containers for security
vulnerabilities. 

## Tags

Note that after the first two releases, version numbers correspond with releases on pypi,
and not Singularity versions.

 - [v0.0.13 (master)](https://github.com/singularityhub/stools) adds support for reports saving
 - [v3.2.1](https://github.com/singularityhub/stools/tree/v3.2.1) Uses Singularity v3.2.1 and above
 - [v2.4.5](https://github.com/singularityhub/stools/tree/v2.4.5) Uses Singularity v2.4.5

## Background
Clair is intended to run as a server to continuous scan Docker *layers* for vulnerabilities. This doesn't map
well to the research domain because of the following:

 - Docker containers come in layers (.tar.gz files) while Singularity images are single binary files that don't "plug in" nicely to Clair.
 - Most researchers can't support continuous running of such a service.

On the other hand, a typical researcher *does* use services like [TravisCI](https://travis-ci.org) and [CircleCI](https://circle-ci.org) to
run their code bases against tests. Since these services now offer running containers and other service-type things, we in fact could do the following:

 - Spin up a Clair server during testing
 - Build a Singularity image, and scan the filesystem contents (before finalized in the image).

While this isn't a continually running service, we can minimally ensure that a container is scanned each time
it is built (and then likely merged to be used in production). If the user takes advantage of [Singularity Hub](https://www.singularity-hub.org) or [Singularity Registry Server](https://singularityhub.github.io/sregistry) the image will be pushed or built for production after passing 
these various tests.

This experiment is based on early discussion in [this thread](https://github.com/singularityhub/sregistry/issues/14).


## Basic Usage

You'll need to first clone the repository:

```bash
git clone https://github.com/singularityhub/stools
cd stools
```

### Build Containers

If you want, build the container (or use a tagged release from [Docker Hub](https://hub.docker.com/repository/registry-1.docker.io/vanessa/stools-clair/tags?page=1)).

```bash
$ docker build -t vanessa/stools-clair .
```

Start the application with [docker compose](https://docs.docker.com/compose/install/). 
Note that you should have the images you want to scan in the $PWD, which will be mapped to the container in `/code` 
(see the [docker-compose.yml](docker-compose.yml) file). You can change this around, just be sure that the containers you want to add are here.

```bash
$ docker-compose up -d
```

Make sure that your containers are up and running! There is one for the clair server
that we will interact with, and one for the database.

```bash
$ docker-compose ps
    Name                   Command               State                            Ports                          
-----------------------------------------------------------------------------------------------------------------
clair-db        docker-entrypoint.sh postgres    Up      0.0.0.0:5432->5432/tcp                                  
clair-scanner   /clair -config=/config/con ...   Up      0.0.0.0:6060->6060/tcp, 6061/tcp, 0.0.0.0:8080->8080/tcp
```

Also note that the folder [reports](reports) by way of being in the mounted present working
directory, will appear at `/code/reports` in the container. We will need to know this later.

### Scan an Image

Let's scan a local image in $PWD mapped to /code in the container. First
pull one from your registry of choice:

```bash
$ singularity pull shub://vsoch/singularity-images
```

And now let's scan! We do this by executing a command to the `clair-scanner` container.
The most basic usage will just print a report to stdout, like this:

```bash
$ docker exec -it clair-scanner sclair singularity-images_latest.sif
...

CVE-2016-9843 (Low)
http://people.ubuntu.com/~ubuntu-security/cve/CVE-2016-9843
The crc32_big function in crc32.c in zlib 1.2.8 might allow context-dependent attackers to have unspecified impact via vectors involving big-endian CRC calculation.
```

### Save a Report

However, if you want to save a report to file (json), you can add the `--report` argument
pointing to an existing output directory of choice. For example, since [reports](reports)
is provided in our present working directory and already bound to the container at `/code/reports`
we can specify that as an argument:

```bash
$ docker exec -it clair-scanner sclair --report /code/reports singularity-images_latest.sif
```

Using `--report` will not disable the print to stdout. However, if you want to disable it,
you can add the `--no-print` option. An example JSON report can be found in the [reports](reports) folder.

```bash
$ docker exec -it clair-scanner sclair --report /code/reports --no-print singularity-images_latest.sif
```

For a full example of what is printed to stdout (using a container with a known vulnerability) see
the [example test](test) folder.


CHANGELOG.md
# CHANGELOG

This is a manually generated log to track changes to the repository for each release.
Each section should include general headers such as **Implemented enhancements** 
and **Merged pull requests**. All closed issued and bug fixes should be 
represented by the pull requests that fixed them. Critical items to know are:

 - renamed commands
 - deprecated / removed commands
 - changed defaults
 - backward incompatible changes (recipe file format? image file format?)
 - migration guidance (how to convert images?)
 - changed behaviour (recipe sections work differently)


## [v1.x](https://github.com/singularityhub/stools/tree/master) (master)
 - bumping aiohttp version (0.0.13)
 - Adding generation of reports to folder with --reports (0.0.12)
 - Original release of stools, primarily will be used with the docker-compose.yml (0.0.1)


stools/clair/__init__.py
#!/usr/bin/env python

"""

Copyright (C) 2018-2021 Vanessa Sochat.

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public
License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""

from stools.version import __version__
from stools.clair.image import export_to_targz
from stools.clair.api import Clair
from stools.clair.server import start
from multiprocessing import Process
import argparse
import os
import requests
import shutil
import sys
import tempfile
import time
import json


def get_parser():
    parser = argparse.ArgumentParser(description="Singularity Clair Scanner")

    parser.add_argument(
        "--version",
        dest="version",
        help="show version and exit.",
        default=False,
        action="store_true",
    )

    parser.add_argument(
        "--start-server",
        dest="server",
        help="If running natively, start the web server too.",
        default=True,
        action="store_true",
    )

    parser.add_argument(
        "images", nargs="*", help="Singularity images to scan.", type=str
    )

    parser.add_argument(
        "--report",
        default=None,
        dest="report_location",
        help="save Clair reports to chosen directory",
    )

    parser.add_argument(
        "--no-print",
        dest="no_print",
        help="Disable printing of report to stdout.",
        default=False,
        action="store_true",
    )

    parser.add_argument(
        "--port",
        default=8080,
        help="port to serve application (default 8080)",
        type=int,
    )

    parser.add_argument(
        "--host",
        default="0.0.0.0",
        help="host to serve application (default 0.0.0.0)",
        type=str,
    )

    parser.add_argument(
        "--clair-port",
        default=6060,
        help="port Clair is running on (default 6060)",
        type=int,
        dest="clair_port",
    )

    parser.add_argument(
        "--clair-host",
        default="0.0.0.0",
        help="host Clair running from (default clair-scanner)",
        type=str,
        dest="clair_host",
    )

    return parser


def version():
    print("\nSingularity Clair Scanner v%s" % __version__)


def main():

    parser = get_parser()

    def help(retval=0):
        """print help, including the software version and active client
        and exit with return code.
        """
        version()
        parser.print_help()
        sys.exit(retval)

    # If the user didn't provide any arguments, show the full help
    if len(sys.argv) == 1:
        help()
    try:
        args = parser.parse_args()
    except:
        sys.exit(0)

    if args.version is True:
        version()
        sys.exit(0)

    # Validate report folder, if provided, exit early if not a directory
    if args.report_location and not os.path.isdir(args.report_location):
        sys.exit("Report directory %s does not exist." % args.report_location)

    # Generate Clair controller
    clair = Clair(args.clair_host, args.clair_port)
    if not clair.ping():
        sys.exit(1)

    print(clair)

    # Local Server
    webroot = "/var/www/images"

    # Start the server and serve static files from root

    if args.server:
        print("\n1. Starting server...")
        server = "http://%s:%s/" % (args.host, args.port)
        process = Process(target=start, args=(args.port, args.host, webroot))
        # start(port=args.port, host=args.host, static_folder=webroot)
        process.daemon = True
        process.start()
        time.sleep(1)

    # Check health of server
    print("\n1. Checking server...")
    response = requests.get(server)
    if response.status_code != 200:
        print("Server not found running at %s" % server)
        sys.exit(1)

    print("2. Processing images!")
    for image in args.images:

        # 1. decompress to sandbox --> tar.gz
        targz = export_to_targz(image)
        print("...exported %s to %s" % (image, targz))

        # 2. Move to webroot
        targz_web = os.path.join(webroot, os.path.basename(targz))
        os.rename(targz, targz_web)
        targz_url = "%simages/%s" % (server, os.path.basename(targz))

        # 3. Scan with Clair, use image name
        print("...serving %s to Clair" % (targz_url))
        clair.scan(targz_web, os.path.basename(image))

        # 4. Generate report
        print("3. Generating report!")
        report = clair.report(os.path.basename(image))
        if args.report_location:
            fpath = os.path.join(
                args.report_location,
                os.path.splitext(os.path.basename(image))[0] + ".json",
            )
            with open(fpath, "w+") as file:
                file.write(json.dumps(report, indent=2))
            print("Wrote report to %s" % fpath)

        # Print to stdout, if desired
        if not args.no_print:
            clair.print(report)

    # Shut down temporary server
    process.terminate()
    shutil.rmtree(webroot)


if __name__ == "__main__":
    main()
